src/main/scala/cognite/spark/v1/FileContentRelation.scala

```diff
@@ -0,0 +1,140 @@
+package cognite.spark.v1
+
+import cats.effect.std.Dispatcher
+import cats.effect.{IO, Resource}
+import com.cognite.sdk.scala.v1.{FileDownloadExternalId, GenericClient}
+import fs2.Stream
+import org.apache.commons.io.FileUtils
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.functions.col
+import org.apache.spark.sql.sources.{BaseRelation, Filter, PrunedFilteredScan, TableScan}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+import org.apache.spark.{Partition, TaskContext}
+import sttp.capabilities.WebSockets
+import sttp.capabilities.fs2.Fs2Streams
+import sttp.client3.asynchttpclient.SttpClientBackendFactory
+import sttp.client3.asynchttpclient.fs2.AsyncHttpClientFs2Backend
+import sttp.client3.{SttpBackend, UriContext, asStreamUnsafe, basicRequest}
+import sttp.model.{Header, Uri}
+
+trait WithSizeLimit {
+  val sizeLimit: Long
+}
+
+class FileContentRelation(config: RelationConfig, fileExternalId: String, inferSchema: Boolean)(
+    override val sqlContext: SQLContext)
+    extends BaseRelation
+    with TableScan
+    with PrunedFilteredScan
+    with Serializable
+    with WithSizeLimit {
+
+  override val sizeLimit: Long = 5 * FileUtils.ONE_GB
+
+  @transient lazy val client: GenericClient[IO] =
+    CdpConnector.clientFromConfig(config)
+
+  @transient private lazy val sttpFileContentStreamingBackendResource
+    : Resource[IO, SttpBackend[IO, Fs2Streams[IO] with WebSockets]] =
+    for {
+      dispatcher <- Dispatcher.parallel[IO]
+      backend <- Resource.make(
+        IO(
+          AsyncHttpClientFs2Backend
+            .usingClient[IO](SttpClientBackendFactory.create("file content download"), dispatcher))
+      )(backend => backend.close())
+    } yield backend
+
+  val acceptedMimeTypes: Seq[String] =
+    Seq("application/jsonlines", "application/x-ndjson", "application/jsonl")
+
+  @transient private lazy val dataFrame: DataFrame = createDataFrame
+
+  override def schema: StructType =
+    dataFrame.schema
+
+  @transient lazy val createDataFrame: DataFrame = {
+    val rdd: RDD[String] = new RDD[String](sqlContext.sparkContext, Nil) with Serializable {
+
+      import cognite.spark.v1.CdpConnector.ioRuntime
+
+      val maxParallelism = 1
+      override def compute(split: Partition, context: TaskContext): Iterator[String] = {
+
+        val validUrl = for {
+          mimeType <- client.files.retrieveByExternalId(fileExternalId).map(_.mimeType)
+          downloadLink <- client.files
+            .downloadLink(FileDownloadExternalId(fileExternalId))
+            .map(_.downloadUrl)
```

Comment:
it may be null? Not marked as required in https://api-docs.cognite.com/20230101/tag/Files/operation/downloadLinks#!c=200&path=items/0/downloadUrl&t=response
but otoh SDK already has some failure handling https://github.com/cognitedata/cognite-sdk-scala/blob/master/src/main/scala/com/cognite/sdk/scala/v1/resources/files.scala#L155
So might be worth checking how downloadlink behaves for not-yet uploaded files (or some other cases without download link?)

Reply:
... there is a "uploaded" on the file api which returns true if it's uploaded. I can just use that I guess.

Reply:
most likely it won't be null and service-contracts just haven't marked it as required, but double-checking with files-api code might still be good

Reply:
No I mean I checked what happens when the file is not uploaded and it's a 400 error. Which seems fair to expose: we first check the "uploaded" status in file api, then if it's set to true, we get the downloadlink here, and if somehow the uploaded is true but the downloadlink is not present (I guess it could happen if the file got deleted in between) we get a 400